Source/ACE/Entity/Player.cs

```diff
@@ -1076,47 +1076,50 @@ public void AddFriend(string friendName)
                 newFriend.Id = new ObjectGuid(friendInfo.Guid, GuidType.Player);
 
                 // Save to DB, assume success
-                DatabaseManager.Shard.AddFriend(Guid.Low, newFriend.Id.Low);
-
-                // Add to character object
-                Character.AddFriend(newFriend);
+                DatabaseManager.Shard.AddFriend(Guid.Low, newFriend.Id.Low, (() =>
+                {
+                    // Add to character object
+                    Character.AddFriend(newFriend);
 
-                // Send packet
-                Session.Network.EnqueueSend(new GameEventFriendsListUpdate(Session, GameEventFriendsListUpdate.FriendsUpdateTypeFlag.FriendAdded, newFriend));
+                    // Send packet
+                    Session.Network.EnqueueSend(new GameEventFriendsListUpdate(Session, GameEventFriendsListUpdate.FriendsUpdateTypeFlag.FriendAdded, newFriend));
+                }));
             }));
         }
 
         /// <summary>
         /// Remove a single friend and update the database.
         /// </summary>
         /// <param name="friendId">The ObjectGuid of the friend that is being removed</param>
-        public async Task<RemoveFriendResult> RemoveFriend(ObjectGuid friendId)
+        public void RemoveFriend(ObjectGuid friendId)
         {
             Friend friendToRemove = Character.Friends.SingleOrDefault(f => f.Id.Low == friendId.Low);
 
             // Not in friend list
             if (friendToRemove == null)
-                return RemoveFriendResult.NotInFriendsList;
+            {
+                ChatPacket.SendServerMessage(Session, "That chracter is not in your friends list!", ChatMessageType.Broadcast);
+                return;
+            }
 
             // Remove from DB
-            DatabaseManager.Shard.DeleteFriend(Guid.Low, friendId.Low);
-
-            // Remove from character object
-            Character.RemoveFriend(friendId.Low);
-
-            // Send packet
-            Session.Network.EnqueueSend(new GameEventFriendsListUpdate(Session, GameEventFriendsListUpdate.FriendsUpdateTypeFlag.FriendRemoved, friendToRemove));
+            DatabaseManager.Shard.DeleteFriend(Guid.Low, friendId.Low, (() =>
+            {
+                // Remove from character object
+                Character.RemoveFriend(friendId.Low);
```

Comment:
I don't believe that the AceCharacter "Friend" functions are thread safe.  Your callback is running from outside of the session's context.  These are likely racy accesses.